### PR TITLE
feat(Quran.com): add presence

### LINKED ISSUE #5891: Quran | quran.com
### Website name

Quran.com

### Website URL

https://quran.com/

### Website logo

you can find it on the website somewhere sorry

### Prerequisites

- [ ] It is a paid service
- [ ] It displays NSFW content
- [ ] It is region restricted

### Description

the quran is said to the be the final revelation from God and the last to come of the abrahamic religions and is the holy book of the muslims

### PR DESCRIPTION
**Pull request by BigBoi-TM**
**Describe the changes in this pull request:**

Resolves #5891
Features:
- Detects homepage
- Detects when a reciter is being read about **_(Eg: "https://quran.com/reciters/161")_**
- Detects when user is on a normal page or if they are reading
#

<details>
<summary> Proof showing the creation/modification is working as expected </summary>

## - Homepage
image

## - Radio
image
`P.S.: The message was changed to "Looking through radio stations..."; this is an older screenshot`

## - Reciter
image
image

## - Reading
image
`P.S.: When you scroll down, everything next to the page number disappears, thus the checks on lines 63-81 and 82-99`

</details>

`Detects a few more pages but they're basic ones and aren't worth taking screenshots of`

### CODE REVIEW

File: websites/Q/Quran.com/presence.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+		injectOnComplete: true
+	}),
+	browsingTimestamp = Math.floor(Date.now() / 1000);
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+			largeImageKey: "logo_1024",
+			startTimestamp: browsingTimestamp
+		},
+		{ pathname } = window.location;
+	switch (true) {
```
Comment (EncryptedDev) [inline websites/Q/Quran.com/presence.ts:12]: all you do in this switch statement is compare a variable to a string so please do a proper switch statement with the variable in here and the cases down below

File: websites/Q/Quran.com/presence.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+				presenceData.details = `Viewing a reciter: ${
+					document.querySelector<HTMLDivElement>(
+						"#__next > div > div.index_pageContainer__Pxtn3 > div.reciterPage_reciterInfoContainer__mYOjC > div > div > div:nth-child(2) > div.ReciterInfo_reciterName__SiK59"
+					).textContent
+				}`;
+			} else presenceData.details = "Looking for a reciter";
+			break;
+		}
+		case pathname === "/about-us": {
+			presenceData.details = "Looking for info about Quran.com";
```
Comment (EncryptedDev) [inline websites/Q/Quran.com/presence.ts:32]:
```suggestion
			presenceData.details = "Viewing the about us page";
```

File: websites/Q/Quran.com/presence.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		}
+		case pathname === "/search": {
+			if (document.title.includes("Search for")) {
+				presenceData.details = `Making a ${document.title
+					.split("-")[0]
+					.toLowerCase()}`;
+			} else presenceData.details = "Searching for something...";
+			break;
+		}
+		case pathname.includes("/juz/" || "/page/"): {
```
Comment (EncryptedDev) [inline websites/Q/Quran.com/presence.ts:63]: `||` does not work here

File: websites/Q/Quran.com/presence.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+				document.querySelector<HTMLSpanElement>(
+					"#__next > div > div.ContextMenu_container__M7_N3 > div > div:nth-child(2) > div > p:nth-child(2) > span"
+				)
+					? document.querySelector<HTMLSpanElement>(
+							"#__next > div > div.ContextMenu_container__M7_N3 > div > div:nth-child(2) > div > p:nth-child(2) > span"
+					  ).textContent
+					: document.querySelector<HTMLSpanElement>(
+							"#__next > div > div.ContextMenu_container__M7_N3.ContextMenu_visibleContainer__KnWDa.ContextMenu_expandedContainer__W_YZP > div > div:nth-child(2) > div > p:nth-child(2) > span.ContextMenu_primaryInfo__QL1fr"
+					  ).textContent
+			})`;
```
Comment (EncryptedDev) [inline websites/Q/Quran.com/presence.ts:79]: Use [nullish coalascing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator)

File: websites/Q/Quran.com/presence.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
```
Comment (EncryptedDev) [inline websites/Q/Quran.com/presence.ts:98]: Use [nullish coalascing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator)

**Review by EncryptedDev — CHANGES_REQUESTED** [review]

File: websites/Q/Quran.com/presence.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
+const presence = new Presence({
+		clientId: "969064871310282813",
+		injectOnComplete: true
```
[the remark this answers is not in the document]
Comment (BigBoi-TM) [reply websites/Q/Quran.com/presence.ts:3]: > Is this really needed?

This was a presence that's been sitting for a good min, I just put it there to see what it did and forgot to take it off. My apologies, will take it off when I get to a computer

File: websites/Q/Quran.com/presence.ts (hunk 4 above)
Comment (BigBoi-TM) [reply websites/Q/Quran.com/presence.ts:79]: Gives me a bunch of errors (when using nullish coalascing), the other version didn't and actually works

Comment (BigBoi-TM) [reply websites/Q/Quran.com/presence.ts:79]: getting errors now cuz the elements have random id's in them, changing to xpaths

Comment (BigBoi-TM) [reply websites/Q/Quran.com/presence.ts:79]: also causes errors using nullish coalascing (either picking up the wrong element cuz somehow the same xPath went to two elements, or gives errors in the console when I flip it), so sticking with this method

File: websites/Q/Quran.com/presence.ts (hunk 3 above)
Comment (BigBoi-TM) [reply websites/Q/Quran.com/presence.ts]: got rid of it completely

File: websites/Q/Quran.com/dist/metadata.json [hunk 7]
```diff
@@ -0,0 +1,23 @@
+{
+	"$schema": "https://schemas.premid.app/metadata/1.7",
+	"author": {
+	  "id": "853932304257908746",
+	  "name": "BigBoi"
+	},
+	"service": "Quran.com",
+	"description": {
+	  "en": "Quran.com is an open source, un-biased, and non-sectarian source of Quranic study that uncovers the beauty of the Quran for Muslims and Non-Muslims. Quran.com offers accessibilty to the Quran translated into many languages in a simple and easy interface."
```
Comment (ImRodry) [inline websites/Q/Quran.com/dist/metadata.json:9]: Unless you tell me where you got this from, I prefer using the official one
```suggestion
	  "en": "The Quran translated into many languages in a simple and easy interface"
```

File: websites/Q/Quran.com/presence.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+			startTimestamp: browsingTimestamp
+		},
+		{ pathname } = window.location;
+	switch (pathname) {
+		case "/": {
+			presenceData.details = "Browsing the homepage...";
+			break;
+		}
+		case "/radio": {
+			presenceData.details = "Looking through radio stations...";
```
Comment (ImRodry) [inline websites/Q/Quran.com/presence.ts:26]:
```suggestion
			presenceData.details = "Looking through radio stations";
```

File: websites/Q/Quran.com/presence.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+		case "/apps": {
+			presenceData.details = "Looking at Quran apps";
+			break;
+		}
+		case "/developers": {
+			presenceData.details = "Looking at the developers page";
+			break;
+		}
+		case "/privacy": {
+			presenceData.details = "Reading privacy terms";
```
Comment (ImRodry) [inline websites/Q/Quran.com/presence.ts:42]: let's not mix things up people
```suggestion
			presenceData.details = "Reading privacy policy";
```

File: websites/Q/Quran.com/presence.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+		}
+		case "/support": {
+			presenceData.details = "Looking at the support page";
+			break;
+		}
+		case "/search": {
+			if (document.title.includes("Search for")) {
+				presenceData.details = `Making a ${document.title
+					.split("-")[0]
+					.toLowerCase()}`;
```
Comment (ImRodry) [inline websites/Q/Quran.com/presence.ts:53]: Gonna be honest I didn't quite understand the "Making a" part but let's do this instead
```suggestion
			const searchQuery = document.querySelector<HTMLinputElement>("#searchQuery")?.value
			if (searchQuery) {
				presenceData.details = `Searching for ${searchQuery}`;
```

File: websites/Q/Quran.com/presence.ts [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+			} else presenceData.details = "Searching for something...";
+			break;
+		}
+		default: {
+			if (pathname.includes("/reciters")) {
+				if (pathname.includes("/reciters/")) {
+					presenceData.details = `Viewing a reciter: ${
+						document.querySelector<HTMLDivElement>(
+							"#__next > div > div.index_pageContainer__Pxtn3 > div.reciterPage_reciterInfoContainer__mYOjC > div > div > div:nth-child(2) > div.ReciterInfo_reciterName__SiK59"
```
Comment (ImRodry) [inline websites/Q/Quran.com/presence.ts:62]:
```suggestion
							".ReciterInfo_reciterName__SiK59"
```

File: websites/Q/Quran.com/presence.ts [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+				presenceData.details = `Making a ${document.title
+					.split("-")[0]
+					.toLowerCase()}`;
+			} else presenceData.details = "Searching for something...";
+			break;
+		}
+		default: {
+			if (pathname.includes("/reciters")) {
+				if (pathname.includes("/reciters/")) {
+					presenceData.details = `Viewing a reciter: ${
```
Comment (ImRodry) [inline websites/Q/Quran.com/presence.ts:60]: The name should go in the state instead

File: websites/Q/Quran.com/presence.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+						document.querySelector<HTMLDivElement>(
+							"#__next > div > div.index_pageContainer__Pxtn3 > div.reciterPage_reciterInfoContainer__mYOjC > div > div > div:nth-child(2) > div.ReciterInfo_reciterName__SiK59"
+						).textContent
+					}`;
+				} else presenceData.details = "Looking for a reciter";
```
Comment (ImRodry) [inline websites/Q/Quran.com/presence.ts:65]:
```suggestion
				} else presenceData.details = "Looking at the reciters";
```

File: websites/Q/Quran.com/presence.ts [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+				} (${
+					getElementByXpath(
+						"/html/body/div[1]/div/div[2]/div/div[2]/div/p[2]/span[2]"
+					)
+						? getElementByXpath(
+								"/html/body/div[1]/div/div[2]/div/div[2]/div/p[2]/span[2]"
+						  ).textContent
+						: getElementByXpath(
+								"/html/body/div[1]/div/div[2]/div/div[2]/div/p[2]/span"
+						  ).textContent
```
Comment (ImRodry) [inline websites/Q/Quran.com/presence.ts:80]: use optional chaining and nullish coalescing

File: websites/Q/Quran.com/presence.ts [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+				})`;
+			}
+		}
+	}
+	presence.setActivity(presenceData);
```
Comment (ImRodry) [inline websites/Q/Quran.com/presence.ts:85]: no check to see if data is present?

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: websites/Q/Quran.com/presence.ts (hunk 14 above)
Comment (BigBoi-TM) [reply websites/Q/Quran.com/presence.ts:80]: image
If you got a way to do it, I'm all ears; in the meantime I'm stuck with this method. Will keep unresolved for your response (hopefully there's one that satisfies you)

File: websites/Q/Quran.com/presence.ts (hunk 15 above)
Comment (BigBoi-TM) [reply websites/Q/Quran.com/presence.ts:85]: There is data right off the bat: The `largeImageKey` and `browsingTimestamp`

File: websites/Q/Quran.com/dist/metadata.json (hunk 7 above)
Comment (BigBoi-TM) [reply websites/Q/Quran.com/dist/metadata.json:9]: Can't remember where I found it but ik for a fact it came from them. Will switch to their default bio for now.

File: websites/Q/Quran.com/presence.ts (hunk 14 above)
Comment (ImRodry) [reply websites/Q/Quran.com/presence.ts:78]: It really isn't that hard
```suggestion
					getElementByXpath(
						"/html/body/div[1]/div/div[2]/div/div[2]/div/p[2]/span[2]"
					) ??
					getElementByXpath(
								"/html/body/div[1]/div/div[2]/div/div[2]/div/p[2]/span"
						  ).textContent
```

**Review by ImRodry — APPROVED** [review]
Besides the unresolved conversation LGTM

**Review by EncryptedDev — APPROVED** [review]

**Comment by EncryptedDev** [comment]
Thank you for the PR

**Pull request merged**